service/worker/scanner/tasklist/scavenger.go

```diff
@@ -0,0 +1,222 @@
+	"time"
+
+	"sync"
+
+	"sync/atomic"
+
+	"github.com/uber-common/bark"
+	"github.com/uber/cadence/common/logging"
+	"github.com/uber/cadence/common/metrics"
+	p "github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/service/worker/scanner/executor"
+)
+
+type (
+	// Scavenger is the type that holds the state for task list scavenger daemon
+	Scavenger struct {
+		db        p.TaskManager
+		executor  executor.Executor
+		metrics   metrics.Client
+		logger    bark.Logger
+		stats     stats
+		startOnce sync.Once
+		stopOnce  sync.Once
+		running   int32
+		stopC     chan struct{}
+		stopWG    sync.WaitGroup
+	}
+
+	taskListKey struct {
+		DomainID string
+		Name     string
+		TaskType int
+	}
+
+	taskListState struct {
+		rangeID     int64
+		lastUpdated time.Time
+	}
+
+	stats struct {
+		tasklist struct {
+			nProcessed int64
+			nDeleted   int64
+		}
+		task struct {
+			nProcessed int64
+			nDeleted   int64
+		}
+	}
+
+	// executorTask is a runnable task that adheres to the executor.Task interface
+	// for the scavenger, each of this task processes a single task list
+	executorTask struct {
+		taskListKey
+		taskListState
+		scvg *Scavenger
+	}
+)
+
+var (
+	taskListBatchSize    = 32                // maximum number of task list we process concurrently
+	taskBatchSize        = 16                // number of tasks we read from persistence in one call
+	maxTasksPerJob       = 256               // maximum number of tasks we process for a executorTask list as part of a single job
+	nWorkers             = taskListBatchSize // number of go routines processing executorTask lists
+	taskListGracePeriod  = 48 * time.Hour    // amount of time a executorTask list has to be idle before it becomes a candidate for deletion
+	epochStartTime       = time.Unix(0, 0)
+	executorPollInterval = time.Minute
+)
+
+// NewScavenger returns an instance of executorTask list scavenger daemon
+// The Scavenger can be started by calling the Start() method on the
+// returned object. Calling the Start() method will result in one
+// complete iteration over all of the task lists in the system. For
+// each task list, the scavenger will attempt
+//  - deletion of expired tasks in the task lists
+//  - deletion of task list itself, if there are no tasks and the task list hasn't been updated for a grace period
+//
+// The scavenger will retry on all persistence errors infinitely and will only stop under
+// two conditions
+//  - either all task lists are processed successfully (or)
+//  - Stop() method is called to stop the scavenger
+func NewScavenger(db p.TaskManager, metrics metrics.Client, logger bark.Logger) *Scavenger {
+	stopC := make(chan struct{})
+	return &Scavenger{
+		db:       db,
+		metrics:  metrics,
+		logger:   logger,
+		stopC:    stopC,
+		executor: executor.NewFixedSizePoolExecutor(taskListBatchSize),
+	}
+}
+
+// Start starts the scavenger
+func (s *Scavenger) Start() {
+	s.startOnce.Do(func() {
+		s.logger.Info("Tasklist scavenger starting")
```

Comment:
nit: maybe some newlines here to differentiate the logic from the metrics etc..
```
		s.logger.Info("Tasklist scavenger starting")

		atomic.StoreInt32(&s.running, 1)
		s.stopWG.Add(1)
		s.executor.Start()
		go s.run()

		s.metrics.IncCounter(metrics.TaskListScavengerScope, metrics.StartedCount)
		s.logger.Info("Tasklist scavenger started")
```

Reply:
I will fix this, but in general, this is something that's going to be pretty much impossible to enforce across the code base - given that we don't have tools today that do this for us - in addition, the rules are unclear. IMO, not worth going after

Reply:
Of course. I don't think this is enforceable or even should be, just something I felt while reading the code - that's why marked as a `nit` - your call whether to fix or not..